src/pip/_internal/index/collector.py

```diff
@@ -350,7 +350,7 @@ def parse_links(page: "HTMLPage", use_deprecated_html5lib: bool) -> Iterable[Lin
     # requested to use html5lib.
     if not use_deprecated_html5lib:
         expected_doctype = "<!doctype html>".encode(encoding)
-        actual_start = page.content[: len(expected_doctype)]
+        actual_start = page.content.lstrip()[: len(expected_doctype)]
```

Comment:
From @pradyunsg's comment:

> ideally without creating a duplicate string of the entire HTML.

Here, `lstrip` is copying the entire HTML. This can be an issue, as (for example) the PyPI index page is huge.

Reply:
I have rewritten it in 3d2984ca6a9be4dbac617fbc312080d7151d83eb so it counts the leading space instead.